filebeat/tests/integration/store_test.go

```diff
@@ -0,0 +1,145 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build integration
+
+package integration
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/elastic/beats/v7/libbeat/tests/integration"
+)
+
+var testStoreCfg = `
+filebeat.inputs:
+  - type: filestream
+    id: test-clean-removed
+    enabled: true
+    clean_removed: true
+    close.on_state_change.inactive: 8s
+    ignore_older: 9s
+    prospector.scanner.check_interval: 1s
+    paths:
+      - %s
+
+filebeat.registry:
+  cleanup_interval: 5s
+  flush: 1s
+
+queue.mem:
+  flush.min_events: 8
+  flush.timeout: 0.1s
+
+path.home: %s
+
+output.file:
+  path: ${path.home}
+  filename: "output-file"
+  rotate_every_kb: 10000
+
+logging:
+  level: debug
+  selectors:
+    - input
+    - input.filestream
+`
+
+func TestStore(t *testing.T) {
+	numLogFiles := 10
+	filebeat := integration.NewBeat(
+		t,
+		"filebeat",
+		"../../filebeat.test",
+	)
+	tempDir := filebeat.TempDir()
+
+	// 1. Create some log files and write data to them
+	logsFolder := filepath.Join(tempDir, "logs")
+	if err := os.MkdirAll(logsFolder, 0755); err != nil {
+		t.Fatalf("could not create logs folder '%s': %s", logsFolder, err)
+	}
+
+	for i := 0; i < numLogFiles; i++ {
+		logFile := path.Join(logsFolder, fmt.Sprintf("log-%d.log", i))
+		integration.GenerateLogFile(t, logFile, 10, false)
+	}
+	logsFolderGlob := filepath.Join(logsFolder, "*")
+	filebeat.WriteConfigFile(fmt.Sprintf(testStoreCfg, logsFolderGlob, tempDir))
+
+	// 2. Ingest the file and stop Filebeat
+	filebeat.Start()
+
+	for i := 0; i < numLogFiles; i++ {
+		// Files can be ingested out of order, so we cannot specify their path.
+		// There will be more than one log line per file, but that at least gives us
+		// some assurance the files were read
+		filebeat.WaitForLogs("Closing reader of filestream", 30*time.Second, "Filebeat did not finish reading the log file")
+	}
+
+	// 3. Remove files so their state can be cleaned
+	if err := os.RemoveAll(logsFolder); err != nil {
+		t.Fatalf("could not remove logs folder '%s': %s", logsFolder, err)
+	}
+	filebeat.WaitForLogs(fmt.Sprintf("%d entries removed", numLogFiles), 30*time.Second, "store entries not removed")
+	filebeat.Stop()
+
+	registryLogFile := filepath.Join(tempDir, "data/registry/filebeat/log.json")
+	readFilestreamRegistryLog(t, registryLogFile, "remove", 10)
```

Comment:
Interesting, I thought we never used `remove` but just updated the `TTL` field.

Reply:
Me too and that was the bug! :exploding_head: 

The resources were never correctly released, so the [`gcClean`](https://github.com/elastic/beats/blob/9fc21609a40d2673998771f5a00d714b279cb8a6/filebeat/input/filestream/internal/input-logfile/clean.go#L102-L110) would never for any entry.

They way Filestream registry works is:
* Whenever an entry is "removed", its TTL is set to zero.
* Periodically the `gcStore` runs, any resource that is finished is then removed from the in-memory store and have a `remove` operation added to the log file.

When [`Find`](https://github.com/elastic/beats/blob/9fc21609a40d2673998771f5a00d714b279cb8a6/filebeat/input/filestream/internal/input-logfile/store.go#L424) is called, it locks the resource by calling `resource.Retain` https://github.com/elastic/beats/blob/9fc21609a40d2673998771f5a00d714b279cb8a6/filebeat/input/filestream/internal/input-logfile/store.go#L428-L431, it's the caller's responsibilit to release it by calling `Release` https://github.com/elastic/beats/blob/9fc21609a40d2673998771f5a00d714b279cb8a6/filebeat/input/filestream/internal/input-logfile/store.go#L465

There were a few places never calling `Release` :man_facepalming:, hence `Finished` https://github.com/elastic/beats/blob/9fc21609a40d2673998771f5a00d714b279cb8a6/filebeat/input/filestream/internal/input-logfile/store.go#L474 **always** returned false and nothing was ever removed from the registry. Which made Filebeat never add the `op: remove` to the registry log file. :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: